mmtk/src/jtoc_call.rs

```diff
@@ -42,27 +47,136 @@ macro_rules! jikesrvm_call {
+    //      : "volatile");
+
+    // The clobber list of the old llvm_asm! is:
+    // "ebx", "ecx", "edx", "esi", "memory", "mm0", "mm1", "mm2", "mm3", "mm4", "mm5", "mm6", "mm7", "xmm0", "xmm1", "xmm2", "xmm3", "xmm4", "xmm5", "xmm6", "xmm7", "ymm0", "ymm1", "ymm2", "ymm3", "ymm4", "ymm5", "ymm6", "ymm7", "zmm0", "zmm1", "zmm2", "zmm3", "zmm4", "zmm5", "zmm6", "zmm7"
+    // When we move to the new asm!, most of the registers are remained as clobbered, except:
+    // * esi: new asm! does not allow use esi as operand (https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/asm/x86.rs#L182), we have to manually save/restore it (following the pattern here: https://github.com/rust-lang/rust/pull/84658/files#diff-d7283132d97a993fad4e2d491ac883dbce4e17fe248cdf37fa3f9334e2a5a115)
+    // * memory: llvm_asm! allows to specify "memory" as clobber if the asm modifies memory.
+    //   asm! has a option nomem to specify the asm does not modify memory, otherwise the compiler assumes the asm will modify memory.
+    // * xmm/ymm: Those are lower parts of zmm. We only need to mark zmm as clobbers.
+
     ($ret:ident, $rvm_thread:ident, $call_addr:ident) => (
-        llvm_asm!("call *%ebx"
-             : "={eax}"($ret)
-             : "{esi}"($rvm_thread), "{ebx}"($call_addr)
-             : "ebx", "ecx", "edx", "esi", "memory", "mm0", "mm1", "mm2", "mm3", "mm4", "mm5", "mm6", "mm7", "xmm0", "xmm1", "xmm2", "xmm3", "xmm4", "xmm5", "xmm6", "xmm7", "ymm0", "ymm1", "ymm2", "ymm3", "ymm4", "ymm5", "ymm6", "ymm7", "zmm0", "zmm1", "zmm2", "zmm3", "zmm4", "zmm5", "zmm6", "zmm7"
-             : "volatile");
+        asm!(
+            // Exchange the value of TLS and esi (so esi now holds the TLS, and the TLS temp holds the esi value)
+            "xchg {0}, esi",
+            // Call $call_addr which is in ebx
+            "call ebx",
+            // Restore esi from the TSL
+            "mov esi, {0}",
+
+            // TLS. We will manually exchange it with esi.
+            inout(reg) $rvm_thread => _,
+            // Call address. We use ebx in asm.
+            inout("ebx") $call_addr => _,
+            // Return value in eax.
+            out("eax") $ret,
+            // Rest of the clobber list.
+            out("ecx") _,
+            out("edx") _,
+            out("mm0") _,
+            out("mm1") _,
+            out("mm2") _,
+            out("mm3") _,
+            out("mm4") _,
+            out("mm5") _,
+            out("mm6") _,
+            out("mm7") _,
+            out("zmm0") _,
+            out("zmm1") _,
+            out("zmm2") _,
+            out("zmm3") _,
+            out("zmm4") _,
+            out("zmm5") _,
+            out("zmm6") _,
+            out("zmm7") _,
+            // equivalent of "volatile" in llvm_asm!
```

Comment:
This doesn't seem right. `nostack` only allows the Rust compiler to make more assumptions (i.e. doesn't use what's below the stack pointer).

[This](https://doc.rust-lang.org/beta/unstable-book/library-features/asm.html) says
> By default, an inline assembly block is treated the same way as an external FFI function call with a custom calling convention: it may read/write memory, have observable side effects, etc. However, in many cases it is desirable to give the compiler more information about what the assembly code is actually doing so that it can optimize better.

So the default is actually what we want, and `nostack` is wrong in this case because the stack can grower to the lower address.